### ISSUE TITLE
Incorrect result when using NULL IN empty subquery

### ISSUE THREAD

**Issue by geraint0923** [prestodb/presto#6622]
```
presto:tiny> select x from (values null, 1) t(x) where (x in (select * from (values 1) p(y) where y < 0)) is null;
  x
------
 NULL
(1 row)
```
The query is expected to return an empty result set.

**Comment by AlekseiS** [comment]
The issue boils down to a question of what should be returned for `NULL in (<empty>)`. Depending on the answer it'll be a bug or not.
Here's the comparison of MySQL and Presto for related cases (the first result is for MySQL, the second is for Presto):
```
select null in (select 1 where false); -- 0/null
select (null in (select 1 where false)) is null; -- 0/true
```
However:
```
select (null in (select 1 where false)) = 0; -- 1/type error
```
As you see, for `NULL in (<empty>)` Presto returns `NULL`, but MySQL returns `0`, so the difference comes from that behavior.

@martint , what do you think is the behavior Presto should have?

**Comment by electrum** [comment]
PostgreSQL also returns false. That behavior sense given that nothing can be in the empty set (not even an unknown value).

**Comment by martint** [comment]
@electrum has it right. It should return false.

**Comment by ArturGajowy** [comment]
# Problem description

The issue stems from how `HashSemiJoinOperator` handles empty tables, or - more precisely - from the fact that it doesn't and can't so far differentiate between empty and non-empty build tables. The relevant part of code is here:

```
for (int position = 0; position < page.getPositionCount(); position++) {
    if (probeJoinPage.getBlock(0).isNull(position)) { 
        blockBuilder.appendNull();    //<-- this is correct only for non-empty build tables
    }
    else {
        boolean contains = channelSet.contains(position, probeJoinPage);
        if (!contains && channelSet.containsNull()) {   //<-- this is correct only as far as SemiJoin's build table is distributed with REPLICATE NULLs
            blockBuilder.appendNull();       
        }
        else {
            BOOLEAN.writeBoolean(blockBuilder, contains);
        }
    }
}
```

The place to differentiate on build table emptiness would be line 3 in the above snippet. It may seem that we can use `channelSet.size() == 0` as an empty table indicator. This does not work for two reasons:

## Predicate pushdown interactions

For simple probe tables, such as in `SELECT NULL IN (...)`, PredicatePushDown is smart enough to push down an always-false filter to the build table, resulting in an empty relation being fetched to the SemiJoin operator. See the query plan:

```
presto:tiny> explain SELECT NULL IN (SELECT 1);
                                                       Query Plan
-------------------------------------------------------------------------------------------------------------------------
 - Output[_col0] => [expr_5:boolean]
         _col0 := expr_5
     - RemoteExchange[GATHER] => expr_5:boolean
         - Project => [expr_5:boolean]
             - SemiJoin[expr = expr_0][$hashvalue, $hashvalue_12] => [expr:integer, $hashvalue:bigint, expr_5:boolean]
                 - LocalExchange[ROUND_ROBIN] () => expr:integer, $hashvalue:bigint
                     - RemoteExchange[REPARTITION][$hashvalue_10] => expr:integer, $hashvalue_10:bigint
                         - Project => [expr:integer, $hashvalue_11:bigint]
                                 $hashvalue_11 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("expr"), 0))
                             - Project => [expr:integer]
                                     expr := CAST(null AS integer)
                                 - LocalExchange[ROUND_ROBIN] () =>
                                     - Values => []
                                             ()
                 - RemoteExchange[REPARTITION - REPLICATE NULLS][$hashvalue_12] => expr_0:integer, $hashvalue_12:bigint
                     - Project => [expr_0:integer, $hashvalue_13:bigint]
                             $hashvalue_13 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("expr_0"), 0))
                         - Project => [expr_0:integer]
                                 expr_0 := 1
                             - Filter[(CAST(null AS boolean) AND ((1 = CAST(null AS integer)) OR (1 IS NULL)))] => []
                                 - LocalExchange[ROUND_ROBIN] () =>
                                     - Values => []
                                             ()

(1 row)
```

## Distribution interactions

Even after disabling the predicate pushdown optimizer, the SemiJoinOperator can't guess if build table is empty because of data distribution. Most of the time (always?) NULLs will be sent to a different node than most (all?) non-null values. For such a query, the SemiJoinOperator will be handed no build-values on the node where the NULL is probe-d. 

# Solution proposals

## Make `SemiJoinNode` require an additional `build_is_empty` input symbol

and base emptiness differentiation on that value. 

SemiJoinNode would then have 3 inputs: `probe`, `build`, and `build_is_empty`. The two latter would use the same subtree, with the difference that predicate pushdown would not be done for `build_is_empty` and that it would wrap the original build subtree in `count(*)` and `limit 1`.

The `build_is_empty` value computation would be planned with the SemiJoinNode itself. Computing the value for the symbol should be quick (it’s LIMIT 1 after all). One concern might be adding another execution stage (a short one, but still).


## Make `channelSet.size() == 0` effectively mean 'the build table is empty'


(seems more cumbersome and hacky)

For this to happen, **both of the following** are needed:

1. Limit the selectiveness of predicates pushed-down to the build table so that it never results in an effectively-empty build table. This can be done in 2 ways:
  a. Disable all predicate pushdown for build tree
  b. Add a function call to the pushed down predicate that will evaluate to true only once on each node. PredicatePushDown would have to add an `OR produce_true_once()` disjunct to the predicates pushed down to SemiJoin's build table as in @sopel39’s [fix proposal](https://github.com/prestodb/presto/compare/master...Teradata:feature_fix_null_in). Note that this means we add a predicate that makes further pushdown in build subtree impossible, because equality inference is no longer possible.
2. A data distribution strategy would have to be created making sure that each node receives at least one build row if there are any. Say, "HASH - REPLICATE NULLS - ALL GET FIRST ROW"

We're curious what you think about that :)

cc @martint @electrum @dain

**Comment by martint** [comment]
The correctness issue is now "fixed" (queries now fail, instead). We should open a separate issue to track adding support for this feature.
